drivers:adc:ad4858 Added ad4858 driver support

mphalke opened the pull request:
Initial implementation of the ad4858 no-os drivers

amiclaus commented on drivers/adc/ad4858/ad4858.c:
```diff
@@ -0,0 +1,1130 @@
+ * @brief Check big endianess of the host processor.
+ * @return Big endianess status (true/false)
+ */
+static bool _is_big_endian(void)
```
check please if you can't use already some functions available in https://github.com/analogdevicesinc/no-OS/blob/master/util/no_os_util.c or https://github.com/analogdevicesinc/no-OS/blob/master/util/

Otherwise add these generic functions (_is_big_endian and _memswap64) there in a separate commit prior to the driver addition.

mphalke replied:
I didn't find the function in util.c that does what I intended- basically wanted to swap the memory content at byte level and also depending upon the step level (e.g. how many bytes I want to swap). Also no function to check the endianness as well. Just to clarify, I have taken these functions from existing ad4170 drivers (developed by Darius) which is present there on our internal bitbucket repo.
Just to be clear, do you suggest to add two separate commits for this? first commit with only these two functions and 2nd with other driver functions?

amiclaus replied:
sorry i wasn't explicit enough. the functions for memswap and is_big_endian should be placed in the no_os_util.c file prior to the driver addition. these are generic functions that should be implemented only once and should not be dependent on a specific part driver.

amiclaus reviewed (commented):
- one inline comment
- some build checks are failing
- i'd suggest dropping extra `_regs`  header/source files and integrate the registers in the main driver header file as macro definitions.

mphalke commented:
> * one inline comment
> * some build checks are failing
> * i'd suggest dropping extra `_regs`  header/source files and integrate the registers in the main driver header file as macro definitions.

-Will fix the build issues
-On 3rd point, this was just added to print the register contents in debug mode. If you see, already these register macros are part of the drivers main header file. What I have done is added them into an array and I am using that array to print the register addresses along with data (done in the init function). This is mainly needed for our application, so if you okay, I can remove these two files completely and remove the debug code added to print  the register contents.

mphalke commented:
V1.0: Review comments fixes

1. Created two commits for the changes. First commit contains additional memswap and endianess check function. 2nd commit contains entire drivers implementation
2. Removed the ad4858_regs.c/.h file as they were added for register debug porpose.
3. Fixed doxygen build errors

amiclaus reviewed (commented):
please add memswap and is_big_endian in no_os_util source files.

amiclaus commented on drivers/adc/ad4858/ad4858.h:
```diff
@@ -0,0 +1,407 @@
+#define AD4858_REG_INTERFACE_CONFIG_A		(AD4858_R1B | 0x00)
+#define AD4858_REG_INTERFACE_CONFIG_B		(AD4858_R1B | 0x01)
+#define AD4858_REG_DEVICE_CONFIG			(AD4858_R1B | 0x02)
+#define AD4858_REG_CHIP_TYPE				(AD4858_R1B | 0x03)
```
the indentation is messed up. check the raw file: https://raw.githubusercontent.com/analogdevicesinc/no-OS/82587278b4871ab99907fcb3f088bfa8c1552b6d/drivers/adc/ad4858/ad4858.h

mphalke replied:
I am using VS and I see alignment properly. Is there a way to check the alignments in raw format on github and adjust accordingly?

![image](https://github.com/analogdevicesinc/no-OS/assets/62383520/b27d1cb1-8241-4a6a-b98b-0c6333abafda)

amiclaus replied:
check all the defines. for some of them I can see that there is a mix of spaces and tabs. for the example you provided here the indentation is correct.

amiclaus commented on drivers/adc/ad4858/ad4858.h:
```diff
@@ -0,0 +1,407 @@
+ */
+enum ad4858_operating_mode {
+	AD4858_NORMAL_OP_MODE,
+	AD4858_LOW_POWER_OP_MODE=0x3,
```
if one member of the enum is explicitly initialized, then all of them should be.

amiclaus commented on drivers/adc/ad4858/ad4858.c:
```diff
@@ -0,0 +1,1069 @@
+	uint32_t addr;
+	uint8_t cnt = 0;
+	uint8_t reg_size = AD4858_LEN(reg_addr);
+	uint8_t buf[6] = { 0 };
```
nit: `{0}`

amiclaus commented on drivers/adc/ad4858/ad4858.c:
```diff
@@ -0,0 +1,1069 @@
+	if (!dev)
+		return -EINVAL;
+
+	if (!dev->addr_ascension_enable) {
```
drop the brackets.

amiclaus commented on drivers/adc/ad4858/ad4858.c:
```diff
@@ -0,0 +1,1069 @@
+		addr = AD4858_ADDR(reg_addr);
+	}
+
+	buf[cnt++] = (uint8_t)(addr >> 8);
```
use no_os_field_get

mphalke replied:
do you mean no_os_put_unaligned_be16? This is the only relevant function I see.
However, the problem with this function is, it's name indicates "be- big endian" but my processor is little endian (as most ARM processors are LE). So this might create some confusion.

amiclaus commented on drivers/adc/ad4858/ad4858.c:
```diff
@@ -0,0 +1,1069 @@
+	}
+
+	buf[cnt++] = (uint8_t)(addr >> 8);
+	buf[cnt++] = (uint8_t)addr;
```
ditto.

amiclaus commented on drivers/adc/ad4858/ad4858.c:
```diff
@@ -0,0 +1,1069 @@
+	if (!dev || !reg_val)
+		return -EINVAL;
+
+	if (!dev->addr_ascension_enable) {
```
drop brackets.

amiclaus commented on drivers/adc/ad4858/ad4858.c:
```diff
@@ -0,0 +1,1069 @@
+	}
+
+	/* Store the address (16-bit instruction phase with R/W=1) */
+	buf[cnt++] = (uint8_t)(AD4858_REG_RD_BIT_MSK | (addr >> 8));
```
no_os_field_get

mphalke replied:
do you mean no_os_put_unaligned_be16?

amiclaus commented on drivers/adc/ad4858/ad4858.c:
```diff
@@ -0,0 +1,1069 @@
+	if (ret)
+		return ret;
+
+	*reg_val = 0;
```
is this explicit initialization required?

mphalke replied:
No I guess, will remove this

amiclaus commented on drivers/adc/ad4858/ad4858.c:
```diff
@@ -0,0 +1,1069 @@
+		return -ETIMEDOUT;
+
+	/* Read SPI data */
+	ret = no_os_spi_write_and_read(dev->spi_desc, buff, nb_bytes);
```
return directly.

amiclaus commented on drivers/adc/ad4858/ad4858.c:
```diff
@@ -0,0 +1,1069 @@
+	if (ret)
+		return ret;
+
+	ret = no_os_gpio_remove(dev->gpio_busy);
```
return directly.

mphalke commented on drivers/adc/ad4858/ad4858.c:
```diff
@@ -0,0 +1,1069 @@
+{
+	int ret;
+	uint8_t indx;
+	uint8_t buff[32] = {};
```
This is just done to send 0 over MOSI line in the SPI data read command as few devices expects 0 over line for data read. However, verified with ad4858 and it doesn't need 0 over MOSI line, so will remove this explicit initialization

amiclaus replied:
then `uint8_t buff[32] = {0};` would be more explicit.

mphalke commented on drivers/adc/ad4858/ad4858.c:
```diff
@@ -0,0 +1,1069 @@
+		return ret;
+
+	if (dev->gpio_busy) {
+		ret = no_os_gpio_direction_input(dev->gpio_busy);
```
I think we should not return as this is optional gpio and if not initialized in init structure, then this if condition won't execute. In that case, we must come to else part and return 0. Correct me if I am wrong.

amiclaus replied:
by return directly I did not mean removing the `return 0;` at the end of the function. Just remove
```
if (ret)
	return ret;
```
and do here:
`return no_os_gpio_direction_input(dev->gpio_busy)`

mphalke commented:
V2.0: Review comments fixes and updated to data capture interface functions

1. Fixed the review comments
2. Fixed issues with data interface logic

amiclaus commented on drivers/adc/ad4858/ad4858.h:
```diff
@@ -0,0 +1,419 @@
+};
+
+/**
+ * @enum ad4858_conv_data
```
struct

mphalke commented:
V3.0; Fixed typo
-Fixed typo in doxygen header for ad4858_conv_data structure
